src/parser/xasl_generation.c

```diff
@@ -2931,6 +3000,113 @@ pt_split_hash_attrs (PARSER_CONTEXT * parser, TABLE_INFO * table_info, PT_NODE *
+ *    O     |        X           | probe attr
+ *    X     |        O           | build attr
+ *    X     |        X           | constant (can be probe or build attr)
+ */
+static int
+pt_split_hash_attrs_for_HQ (PARSER_CONTEXT * parser, PT_NODE * pred, PT_NODE ** build_attrs, PT_NODE ** probe_attrs)
+{
+  PT_NODE *tmp = NULL, *pointer = NULL;
+  PT_NODE *build_nodes = NULL;
+  PT_NODE *probe_nodes = NULL;
+  int cur_build, num_attrs, i;
+  PT_NODE *attr_list = NULL;
+  PT_NODE *node = NULL, *save_node = NULL, *save_next = NULL;
+  PT_NODE *arg1, *arg2;
+
+  assert (build_attrs != NULL && *build_attrs == NULL);
+  assert (probe_attrs != NULL && *probe_attrs == NULL);
+  *build_attrs = NULL;
+  *probe_attrs = NULL;
+
+  if (pred)
+    {
+      /* Traverse pred */
+      for (node = pred; node; node = node->next)
+	{
+	  save_node = node;	/* save */
+
+	  CAST_POINTER_TO_NODE (node);
+
+	  if (!PT_IS_EXPR_NODE_WITH_OPERATOR(node, PT_EQ) || node->or_next)
+	    {
+	      /* HASH LIST SCAN for HQ is possible under the following conditions */
+	      /* 1. CNF predicate (node is NOT PT_AND, PT_OR) */
+	      /* 2. only equal operation */
+	      /* 3. predicate without OR (or_next is null) */
+	      /* 4. symmetric predicate (having PRIOR, probe. having NAME, build. Having these two makes it unhashable) */
+	      /* 5. subquery is not allowed in syntax check */
+	      continue;
+	    }
+	  else
+	    {
+	      /* save and cut-off node link */
+	      save_next = node->next;
+	      node->next = NULL;
+
+	      arg1 = node->info.expr.arg1;
+	      arg2 = node->info.expr.arg2;
+
+	      HASHABLE hashable_arg1, hashable_arg2;
+	      hashable_arg1 = hashable_arg2 = {false, false};
+	      HASH_ATTR hash_arg1, hash_arg2;
+
+	      parser_walk_tree (parser, arg1, pt_check_hashable, &hashable_arg1, NULL, NULL);
+	      parser_walk_tree (parser, arg2, pt_check_hashable, &hashable_arg2, NULL, NULL);
+
+	      CHECK_HASH_ATTR (hashable_arg1, hash_arg1);
+	      CHECK_HASH_ATTR (hashable_arg2, hash_arg2);
+
+	      if ((hash_arg1 == PROBE && (hash_arg2 == BUILD || hash_arg2 == CONSTANT)) ||
+		  (hash_arg1 == CONSTANT && hash_arg2 == BUILD))
+		{
+		  /* arg1 is probe attr */
+		  *build_attrs = parser_append_node (parser_copy_tree (parser, arg2), *build_attrs);
+		  *probe_attrs = parser_append_node (parser_copy_tree (parser, arg1), *probe_attrs);
+		}
+	      else if ((hash_arg1 == BUILD && (hash_arg2 == PROBE || hash_arg2 == CONSTANT)) ||
+		       (hash_arg1 == CONSTANT && hash_arg2 == PROBE))
+		{
+		  /* arg1 is build attr */
+		  *build_attrs = parser_append_node (parser_copy_tree (parser, arg1), *build_attrs);
+		  *probe_attrs = parser_append_node (parser_copy_tree (parser, arg2), *probe_attrs);
+		}
+	      else
+		{
+		  /* unhashable predicate */
+		}
+
+	      /* restore node link */
+	      node->next = save_next;
+	    }
+
+	  node = save_node;	/* restore */
+	}			/* for (node = ...) */
+    }
+
+  return NO_ERROR;
+
+exit_on_error:
```

Comment:
it can't reach in this label's block. Check if you omitted to check er_has_error() somewhere.

Reply:
I can find something related to error. What about putting it in for later use?

Reply:
it ok for only later use.